### ISSUE TITLE
Tracking issue for video decoding

### ISSUE THREAD

**Issue by Dinnerbone** [ruffle-rs/ruffle#723]
We've had a number of people ask about this, and whilst we may not solve it soon - we can use this to let them be updated when we make progress.

There's 4 codecs (that I know of) that we'll need to support inside `DefineVideoStream`:

- [ ] Sorenson H.263 (codec 2)
- [ ] Screen Video (codec 3)
- [ ] VP6 (codec 4)
- [ ] VP6 with Alpha (codec 5)

I can also see that H.264 is required according to https://helpx.adobe.com/flash/kb/supported-codecs-flash-player.html

The biggest limitations will be that we need decoders that satisfy the following requirements:
- Must be usable on desktop on Mac, Windows and Linux
- Must be usable on browser either as built-in to the browser, or through Wasm
- Must be distributable legally
  - Video codecs and patents or licensing terms unfortunately go hand-in-hand
  - Asking end-users to install something would be a no-go for web

Issue #590 contains a relevant discussion on using ffmpeg to solve this.

Issue will be updated with more info as we progress.

**Comment by ROBERT-MCDOWELL** [comment]
about h264 dec and enc, I think the best choice would be ffmpeg lib converted in wasm or
just https://github.com/cisco/openh264
already used by chrome in PPAPI and firefox in NPAPI

**Comment by EZWrighter** [comment]
Adding a note here, that RTMP streaming is still used often in the browser for close to real time video delivery and flash is the only solution.  This would be a huge feature if we could still stream RTMP after flash is no longer officially supported in browsers.

**Comment by ROBERT-MCDOWELL** [comment]
it looks like ffmpeg in wasm is more and more efficient and workers ready, which can open the door to a vast encoding decoding, including all Flash codecs of course
https://github.com/ffmpegwasm/ffmpeg.wasm

**Comment by EZWrighter** [comment]
I actually looked at this problem and I believe socket security is a huge issue for the streaming libraries, so I am guessing RTMP Streaming will never be possible via wasm :-(

**Comment by ROBERT-MCDOWELL** [comment]
this is not correct, Rust has already an RTMP library. 
https://users.rust-lang.org/t/rtmp-live-streaming-in-rust/16869
alos, with QUIC, future HTTP3, TCP and UDP low level protocol is available.

**Comment by EZWrighter** [comment]
But that isn't in a wasm assembly.  From looking at that, it looks like an RTMP server implementation.  The problem is with opening up a socket to communicate with the RTMP server from within the assembly.  I don't think that will be allowed.

**Comment by ROBERT-MCDOWELL** [comment]
rtmp socket is just a tcp socket,
https://github.com/robymus/rtmp.js-websocket
https://github.com/yurydelendik/rtmp.js
https://github.com/illuspas/Node-Media-Server/blob/master/node_rtmp_client.js
 rtmfp needs an udp socket, available for now only with webrtc.
https://hpbn.co/webrtc/

**Comment by EZWrighter** [comment]
Looking at those projects, they all require some one off browser configuration...none of them are general purpose solutions.

My understanding is that the communication has to come from a higher level and then be bridged down into the webassembly. https://github.com/ffmpegwasm/ffmpeg.wasm/issues/6   Maybe I am just out of my league here with the communication paths.

**Comment by kmeisthax** [comment]
All of the links Robert posted are for accessing RTMP streams in non-browser environments. For security reasons, code running in the context of a browser is prohibited from making arbitrary TCP or UDP connections. It is only allowed to make HTTP, WebSocket, and WebRTC requests to it's own origin or a consenting third-party origin using Fetch, XMLHttpRequest, or by including the content in some other high-level tag. There is no way to enforce a same-origin policy with sockets (you need an HTTP `Host` header in order to validate origins), ergo you don't get sockets on the web.

As for ffmpeg.wasm, in order to get H.263 or VP6 support we need to build it ourselves, which requires everyone who wants to build Ruffle to also have a working Emscripten build environment and git clones. This would be harder to integrate with cargo or CI systems. The story around mixed-language projects is a bit murkier on web than on desktop, so I'd prefer having access to pure-Rust software decoders. If I'm completely wrong about this, let me know, but we may ultimately have to write our own decoders for at least some of these codecs.

**Comment by lordhugo7880** [comment]
What is the current status for FLV and F4V video container support? Has this been fully implemented yet?

**Comment by lordhugo7880** [comment]
What is the current status for FLV and F4V video container support? Have these been fully implemented yet?

**Comment by Herschel** [comment]
No video implementation at all currently.

**Comment by lordhugo7880** [comment]
Ok, thanks for the info. Btw, according to page 72 (page 78 of the PDF file) of this document: [http://download.macromedia.com/f4v/video_file_format_spec_v10_1.pdf](url) ,  'Screen video version 2' is listed as a separate video codec that needs to be supported.

**Comment by darkdragon-001** [comment]
It would already help me a lot if `.flv` videos encoded via h.264 could be played using the browser codec. This way, an easy software path could be developed and tested while additional codecs could more easily be added later.

**Comment by torokati44** [comment]
Just hypothetically, being limited to codec 2:
Let's say someone pulls a pure Rust, portable implementation of a Sorenson Spark / FLV1 / H.263 decoder out of a magic hat.
 - Would there be any issues with patents in this case? I haven't found expiration info about this particular format (variant), only about H.264, MPEG4, and similar.
 - In addition to the codec itself, how much more work would be putting in place the rest of the video playback infrastructure? Is it perhaps already mostly there - given that animated GIFs are already supported?

**Comment by kmeisthax** [comment]
*reaches into his magic hat* https://github.com/kmeisthax/ruffle/tree/video-h263

 * Sorenson is more or less H.263 with *less* features. The only thing they added was disposable frames (which we could just retain anyway), a different/wider way of encoding IDCT `LEVEL`s, and a relaxation of the H.263 motion vector rules.
 * My branch is currently stuck on not having an easy/portable way to copy decoded video data into the rendering backend. I was going to special-case each one, but then I saw the `BitmapData` PR and decided to wait for that to get merged in before I pick up work on video again.

**Comment by grishka** [comment]
I have too much experience with low-level multimedia, though not "writing my own video decoder" low-level, but close to that at times. I also am very nostalgic about Flash in general, not least because it literally substantially changed my life around 10 years ago. So, I'd totally help you with this, but the problem is — I've never written a single line of Rust, I'm more of an Android/Java/C++ kind of guy. Looking at the existing code, it seems like there's no infrastructure at all for video, so I've got no idea where to even start.

Oh okay, while I was typing this comment, it turned out someone is already working on H263. Anyway, if you need any help with video that doesn't require knowledge of Rust, I'm all yours 😌

**Comment by torokati44** [comment]
Well, the BitmapData PR is now in master! \o/ Does it make the further progress on the h263 decoder possible, @kmeisthax? Is there something I could possibly help with?

**Comment by kmeisthax** [comment]
BitmapData will be super helpful, I don't have to do downcast shenanigans just to get the video data in.

I'll need to finish the AS3 event dispatch stuff first before I jump back to this, though, and I'm probably going to have multiple cycles of back-and-forth on the video decoder before I can call it ready to merge. It technically has yet to even decode a single Sorenson H.263 bitstream, though it is complete *enough* to have all features necessary to do so. It's just untested.

**Comment by toxtox** [comment]
@kmeisthax 
H.264 codec support would be very helpful.
Since the birth of our children 8 years ago I am running a very huge family picture and video gallery with already several TB. 
The gallery is running with the software http://galleryproject.org/. The flash player is needed to play video files. Since today adobe flash player is end of life :-( . I already installed the **_ruffle_nightly_2021_01_12_firefox.xpi_** extension, the player gui shows up, but the ruffle player cannot play the mp4-video-file because of the missing video codecs.

Maybe the source code of "**_Lightspark_**" https://github.com/lightspark/lightspark could be helpful to add the **H.263** and **H264** and MP3 sound codec because it supports both codecs via ffmpeg https://www.golem.de/1008/77140.html .
![Gallery2_Project_with_ruffle](https://user-images.githubusercontent.com/15896490/104377402-bfce0400-5526-11eb-96fb-fc13ea1b0ec7.png)

**Comment by sombraguerrero** [comment]
Worth noting that Onda Educa is in need of FLV decoding, particularly for their Colección Animacuentos application.
In the attached screenshot, a FLV is supposed to play in the leftmost frame, and the other two are to follow along in sync.
![toto](https://user-images.githubusercontent.com/24618934/111897538-6d113c80-89ee-11eb-862c-ec247e34e238.png)

**Comment by lvyitian** [comment]
reference: https://github.com/bilibili/flv.js

**Comment by ROBERT-MCDOWELL** [comment]
Also I really think the best to implement any kind of audio/video protocols and codecs would be WebTransport (HTTP3)
which include UDP/TCP low level socket and ready to work with all browsers in wasm
https://en.wikipedia.org/wiki/HTTP/3
allowing Flash protocol implementation to be developed on top of it.
this project can also be used for now since webtransport 1.0 in rust is not relased yet
https://github.com/kyren/webrtc-unreliable

**Comment by torokati44** [comment]
With https://github.com/ruffle-rs/ruffle/pull/5369 now in, two more boxes can be checked! \o/

**Comment by torokati44** [comment]
BTW, If anyone comes around (or is willing to help me create) any .swf with a "Screen Video V2" embedded, please hit me up.
I tried really hard to find any examples, but wasn't successful so far.
I also tried to make one by utilizing the encoder in ffmpeg, but it's not willing to mux it into an .swf file, only into a .flv.
And even if I modify ffmpeg so that it is willing to do so, Flash Player doesn't play the result...

**Comment by ROBERT-MCDOWELL** [comment]
did you check the Screen Video V2 specs?
https://www.adobe.com/content/dam/acom/en/devnet/pdf/swf-file-format-spec-v10.pdf

**Comment by torokati44** [comment]
Of course, but thanks!
Although, the format actually understood by Flash Player (in good Adobe fashion) doesn't even match the spec for Screen Video V1 perfectly, so I wouldn't put that much weight on it...
And the encoded V2 video itself _should_ be fine, it's just embedding it into an SWF that's problematic at the moment.

**Comment by torokati44** [comment]
I think #5808 checked another (the last, for now?) box here!

**Comment by darkdragon-001** [comment]
Great work! Any plan on how to proceed with h.264?

**Comment by ROBERT-MCDOWELL** [comment]
another promising rust RTMPx project
https://kalldrexx.github.io/mmids/

**Comment by ROBERT-MCDOWELL** [comment]
FYI SWF specs PDF as the adobe link is broken
https://github.com/blackears/raven/blob/master/proj/SWFParser/doc/swf_file_format_spec_v10.pdf

**Comment by ROBERT-MCDOWELL** [comment]
FYI FLV File Format Specs PDF
https://github.com/NewChromantics/PopCodecs/blob/master/adobe_flv_video_file_format_spec_v10.pdf

**Comment by grishka** [comment]
If I remember correctly, FLV is more or less the receive side an RTMP stream dumped into a file, so some of the code could probably be shared between the two.

**Comment by danielhjacobs** [comment]
H.264 is supported on Desktop now. Support on web is being implemented by https://github.com/ruffle-rs/ruffle/compare/master...torokati44:ruffle:external-video-webcodecs (VideoDecoder) and https://github.com/torokati44/ruffle/compare/external-video-webcodecs...torokati44:ruffle:external-video-mse (MSE fallback).

**Comment by ROBERT-MCDOWELL** [comment]
@torokati44 
about Screen Video V2 codec, maybe if you download the ffmpeg sources
and from ffmpeg/libavcodec/flashsv2 you will get all the C code of the encoder and decoder.

**Comment by lvyitian** [comment]
> @torokati44 about Screen Video V2 codec, maybe if you download the ffmpeg sources and from ffmpeg/libavcodec/flashsv2 you will get all the C code of the encoder and decoder.

@ROBERT-MCDOWELL This violates the LGPLv2.1 License.

**Comment by torokati44** [comment]
> @ROBERT-MCDOWELL This violates the LGPLv2.1 License.

Yepp.

The issue is not with the code. There already is a Rust decoder for Screen V2 in NihAV. I'm sure the author, Kostya, would (or maybe he alread has!) allow us to use/incorporate it, like we did with his V1 decoder.

The reason it's not there yet is that I haven't been able to find a single example of a .swf or .flv file in the wild that uses it, nor any way to make one.

If you can point me to some of these, we can get to it.

**Comment by ROBERT-MCDOWELL** [comment]
@torokati44 sad :-\
but As I said about the swf or flv example you just need to convert a video file to a flash video screen V2 with ffmpeg no?

**Comment by torokati44** [comment]
Yes, that could work. 😅 Maybe my issue was back when we didn't have FLV support, and that it couldn't put it into an SWF? Or something like that...

Either way, until there is no real use case found for it, I'm not personally inclined to get it in just to tick that checkbox. It would be nice, and maybe some day...

**Comment by ROBERT-MCDOWELL** [comment]
I just converted one for you....
[sample-flashSV2.flv.txt](https://github.com/user-attachments/files/15787352/sample-flashSV2.flv.txt)

**Comment by ROBERT-MCDOWELL** [comment]
here is another without audio. Screen Video V2 cannot be muxed into a SWF apparently. only V1 is allowed
[sample-flashSV2.flv.txt](https://github.com/user-attachments/files/15787677/sample-flashSV2.flv.txt)

**Comment by torokati44** [comment]
When substituting either of these files into the .flv of the H.264 test, I get these glitches from time to time:

![image](https://github.com/ruffle-rs/ruffle/assets/288816/f1236afb-af6a-4d83-9fb6-5bf5a50cf560)
![image](https://github.com/ruffle-rs/ruffle/assets/288816/467dc124-eecf-4120-8055-b2740be0c903)
![image](https://github.com/ruffle-rs/ruffle/assets/288816/9421de8a-8b56-415e-b803-503bf292c944)

However, they play perfectly in any other (native) video player.

This hints at both the encoder and decoder in FFmpeg being subtly wrong, cancelling each other in their wrongness.

So, an "actual, real world" .FLV, that plays flawlessly in FP, would still be necessary. Even to fix FFmpeg.

**Comment by ROBERT-MCDOWELL** [comment]
this glitches are maybe related to a FP buffer to set correctly for this special Screen Video V2 codec?

**Comment by torokati44** [comment]
> this glitches are maybe related to a FP buffer to set correctly for this special Screen Video V2 codec? 

I'm not sure what you mean by this... AFAIK the player in that test is a generic FLV video playery and apparently it does recognize the used codec correctly.

> I doubt it's a bug on ffmpeg side, especially when native video players play it well too.

Those players more than likely use the same decoder from libavcodec. :)
What I mean is, that encoder/decoder pair is probably consistent with itself, but may not perfectly match the format expected and understood by FP. Sure, almost, but not exactly.

Could you point me to the tests you mentioned? Do they check FP parity/compatibility?

**Comment by ROBERT-MCDOWELL** [comment]
I cannot test more than you, the thing I know is Screen Video V2 was exclusively used with adobe connect to stream the client desktop and if I remember well had a special "key(?),code(?), payload(?)"  to now allow hackers/users to use this codec out of adobe connect. reason why maybe there are glitches. but I'm not really sure.

**Comment by Dinnerbone** [comment]
The biggest question is about practical impact of delivering this feature: is there content that exists today which uses this in Flash Player, or are we trying to implement something that nobody practically used? We have a lot of things we need to deliver for Ruffle that *are* being used, before chasing down things that only exist for the sake of existing.

If we're strugging to even make a test that uses this, I have a hard time believing that any real world content uses it either.

**Comment by torokati44** [comment]
Yeah, I mostly agree (https://github.com/ruffle-rs/ruffle/issues/723#issuecomment-2160574869).

**Comment by ROBERT-MCDOWELL** [comment]
@Dinnerbone @torokati44  ho yes I totally agree, this is something that can be done at the very end, a lot of other points are more important indeed!
